File: core/vaultcrypto/src/main/kotlin/xyz/libravault/core/vaultcrypto/VaultFileReader.kt
package xyz.libravault.core.vaultcrypto

import java.io.Closeable
import java.io.RandomAccessFile
import java.nio.ByteBuffer

/**
 * Seekable, random-access decrypting reader over a chunked vault file (PRD §8.2).
 *
 * This is the primitive that makes the whole content-delivery architecture work
 * (implementation plan §B/§D.0.RESULTS): a PDF proxy fd, a Media3 [DataSource][
 * androidx.media3.datasource.DataSource]-style adapter, or a [MediaDataSource][
 * android.media.MediaDataSource] can all be built as a thin wrapper around
 * [readAt] — none of them need the whole file decrypted up front. Those Android
 * adapters live in core:vaultstore/feature layers (Phase 3), not here; this
 * class has zero Android dependencies so it's fully JVM-testable.
 *
 * Even a zero-byte file gets exactly one (empty) authenticated chunk written by
 * [ChunkedVaultWriter] — this is deliberate: without it, an attacker could
 * remove *every* chunk from a file and there would be nothing left whose AEAD
 * tag could fail, defeating tamper detection entirely for that case.
 *
 * Caches the most recently decrypted chunk, since both PDF page rendering and
 * audio playback read sequentially within a chunk far more often than they
 * jump across chunk boundaries.
 *
 * **Not thread-safe.** One instance wraps one reusable [AesGcmCipher] (itself
 * not thread-safe — see its doc) plus mutable single-chunk cache state, so
 * concurrent calls to [readAt] on the same instance from multiple threads are
 * unsafe. Open one instance per concurrent reader (e.g. per proxy-fd callback
 * on its own dedicated thread — see implementation plan §D.0.RESULTS); do not
 * share one instance across simultaneously-active callbacks.
 *
 * @param expectedFileId the caller's already-known identity for this blob (e.g. from a
 * manifest entry) — cross-checked against the id embedded in the file's own header, so a
 * caller that opens the wrong blob by mistake gets a fast, clear error instead of an opaque
 * authentication failure three steps later. Pass `null` when the caller has no independent
 * way to know the identity ahead of time (currently only core:vaultstore's manifest, which
 * lives at a single fixed path with nothing else to cross-check against) — the id embedded
 * in the header is then trusted directly to derive the key. This weakens nothing: every header
 * field, fileId included, is bound into every chunk's AEAD tag ([VaultFormat.chunkAad]), so
 * a tampered fileId already fails to decrypt (wrong derived key) with or without this
 * pre-check; the check is purely a nicer failure mode, never the actual security boundary.
 * @throws VaultAuthenticationException wrong VMK, or the file was tampered with
 * @throws VaultTruncatedException the file is shorter than the header requires
 * @throws UnsupportedVaultFormatException the header's format version isn't one this build understands
 * @throws UnsupportedCipherException the header's cipher id isn't one this build understands
 * @throws MalformedVaultHeaderException a header field is structurally invalid (e.g. chunkSize <= 0)
 */
class VaultFileReader(
    file: java.io.File,
    private val vmk: ByteArray,
    expectedFileId: ByteArray?,
) : Closeable {

    private val raf = RandomAccessFile(file, "r")
    private val cipher = AesGcmCipher() // one instance, reused across every chunk — see AesGcmCipher's doc

    val fileId: ByteArray
    val chunkSize: Int
    val plainSize: Long
    private val formatVersion: Byte
    private val cipherId: Byte
    private val fileContentKey: ByteArray

    private var cachedChunkIndex = -1L
    private var cachedChunk: ByteArray? = null

    /** Counts actual chunk decryptions (cache misses) — lets tests assert the
     * cache and lazy random-access behavior actually work as designed, which is
     * the entire premise the Phase 0 spike validated on real hardware. */
    @Volatile var decryptCount: Int = 0
        private set

    init {
        val header = ByteArray(VaultFormat.HEADER_SIZE_BYTES)
        val headerRead = raf.read(header)
        if (headerRead < VaultFormat.HEADER_SIZE_BYTES) {
            raf.close()
            throw VaultTruncatedException(VaultFormat.HEADER_SIZE_BYTES.toLong(), maxOf(headerRead, 0).toLong())
        }
        val buf = ByteBuffer.wrap(header)
        val version = buf.get()
        val cipherId = buf.get()
        val parsedFileId = ByteArray(VaultFormat.FILE_ID_SIZE_BYTES).also { buf.get(it) }
        val parsedChunkSize = buf.int
        val parsedTotalLength = buf.long

        if (version != VaultFormat.FORMAT_VERSION) {
            raf.close()
            throw UnsupportedVaultFormatException(version)
        }
        if (cipherId != VaultFormat.CIPHER_AES_256_GCM) {
            raf.close()
            throw UnsupportedCipherException(cipherId)
        }
        require(expectedFileId == null || expectedFileId.size == VaultFormat.FILE_ID_SIZE_BYTES) {
            "expectedFileId must be ${VaultFormat.FILE_ID_SIZE_BYTES} bytes"
        }
        // null means the caller has no independent way to know the identity ahead of time —
        // trust whatever the header says (see the class doc's @param expectedFileId). Every
        // header field, including fileId, is still bound into every chunk's AEAD tag, so a
        // tampered fileId fails to decrypt regardless of this check ever running.
        if (expectedFileId != null && !expectedFileId.contentEquals(parsedFileId)) {
            raf.close()
            // Not itself an AEAD failure — a fast, clear error for "this is the wrong
            // file's blob" rather than letting the caller discover it as an opaque
            // authentication failure three steps later on the first chunk read.
            throw VaultAuthenticationException(
                IllegalStateException("fileId mismatch: expected != stored header fileId"),
            )
        }
        // Structural validation BEFORE any arithmetic uses these values. Found
        // during PR review: chunkSize == 0 previously reached an unguarded
        // division in chunkCountFor and crashed with an unhandled
        // ArithmeticException instead of failing the same clean way every other
        // corrupted-header case does.
        if (parsedChunkSize <= 0 || parsedChunkSize > VaultFormat.MAX_REASONABLE_CHUNK_SIZE) {
            raf.close()
            throw MalformedVaultHeaderException("Invalid chunkSize in header: $parsedChunkSize")
        }
        if (parsedTotalLength < 0) {
            raf.close()
            throw MalformedVaultHeaderException("Negative totalPlaintextLength in header: $parsedTotalLength")
        }

        this.fileId = parsedFileId
        this.formatVersion = version
        this.cipherId = cipherId
        this.chunkSize = parsedChunkSize
        this.plainSize = parsedTotalLength
        this.fileContentKey = deriveFileContentKey(vmk, parsedFileId)

        // Eagerly authenticate chunk 0 at open time, even for a legitimately empty
        // (plainSize == 0) file. This is deliberate, not incidental: readAt() below
        // never touches disk for an empty file (there is nothing to return), which
        // means without this eager check, an attacker could truncate a NON-empty
        // file down to just the header and rewrite totalPlaintextLength to 0 —
        // "erasing" its content as a silent, plausible-looking empty file instead
        // of a loud failure. Decrypting chunk 0 here closes that gap: it's what
        // makes the "every file has at least one authenticated chunk" guarantee
        // (see this class's doc comment) actually hold for every open, not just
        // for opens that happen to read content.
        decryptChunk(0)
    }

    private val chunkCount: Long get() = ChunkedVaultWriter.chunkCountFor(plainSize, chunkSize)

    private fun decryptChunk(index: Long): ByteArray {
        cachedChunk?.let { if (cachedChunkIndex == index) return it }

        val isFinal = index == chunkCount - 1
        val plainLenOfChunk = if (isFinal) (plainSize - index * chunkSize) else chunkSize.toLong()
        val storedLen = plainLenOfChunk + VaultFormat.TAG_SIZE_BYTES
        val filePos = VaultFormat.HEADER_SIZE_BYTES + index * (chunkSize + VaultFormat.TAG_SIZE_BYTES)

        val ciphertext = ByteArray(storedLen.toInt())
        raf.seek(filePos)
        var offset = 0
        while (offset < ciphertext.size) {
            val n = raf.read(ciphertext, offset, ciphertext.size - offset)
            if (n < 0) throw VaultTruncatedException(ciphertext.size.toLong(), offset.toLong())
            offset += n
        }

        val aad = VaultFormat.chunkAad(formatVersion, cipherId, fileId, plainSize, chunkSize, index, isFinal)
        val nonce = deriveNonce(fileContentKey, index)
        val plaintext = cipher.decrypt(fileContentKey, nonce, aad, ciphertext)

        decryptCount++
        cachedChunkIndex = index
        cachedChunk = plaintext
        return plaintext
    }

    /**
     * Reads up to [length] plaintext bytes starting at plaintext [offset].
     * Returns fewer bytes than requested only at end-of-file (mirrors the
     * contract callers like a [ProxyFileDescriptorCallback][
     * android.os.ProxyFileDescriptorCallback] or [MediaDataSource][
     * android.media.MediaDataSource] expect).
     *
     * @throws VaultAuthenticationException wrong key, or the file was tampered with
     * @throws VaultTruncatedException fewer bytes exist on disk than the header implies
     */
    fun readAt(offset: Long, length: Int): ByteArray {
        if (offset < 0 || length < 0) return ByteArray(0)
        if (offset >= plainSize) return ByteArray(0)

        val want = minOf(length.toLong(), plainSize - offset).toInt()
        val result = ByteArray(want)
        var produced = 0
        var pos = offset
        while (produced < want) {
            val index = pos / chunkSize
            val withinChunk = (pos % chunkSize).toInt()
            val chunk = decryptChunk(index)
            val take = minOf(chunk.size - withinChunk, want - produced)
            if (take <= 0) break
            System.arraycopy(chunk, withinChunk, result, produced, take)
            produced += take
            pos += take
        }
        return if (produced == want) result else result.copyOf(produced)
    }

    override fun close() = raf.close()
}

File: ios/LibraVaultApp/LibraVault/LibraVault/Sources/VaultCrypto/VaultFileReader.swift
import Foundation

/// Seekable, random-access decrypting reader over a chunked vault file (PRD §8.2).
///
/// This is the primitive Android's equivalent doc comment describes as making
/// the whole content-delivery architecture work: a PDF/audio data-source
/// adapter can be built as a thin wrapper around `readAt` - none of them need
/// the whole file decrypted up front. Those adapters live above this module
/// (in whatever iOS equivalent of core:vaultstore/feature ends up wrapping
/// it), not here; this type has zero UIKit/AVFoundation dependency, same as
/// the Kotlin original has zero Android dependency.
///
/// Even a zero-byte file gets exactly one (empty) authenticated chunk written
/// by `ChunkedVaultWriter` - deliberate: without it, an attacker could remove
/// *every* chunk from a file and there would be nothing left whose AEAD tag
/// could fail, defeating tamper detection entirely for that case.
///
/// Caches the most recently decrypted chunk, since both PDF page rendering and
/// audio playback read sequentially within a chunk far more often than they
/// jump across chunk boundaries.
///
/// **Not thread-safe.** Mutable single-chunk cache state plus one `FileHandle`
/// - open one instance per concurrent reader; do not share one instance across
/// simultaneously-active callers.
///
/// - Throws (from `init`): `.authenticationFailed` (wrong VMK, or the file was tampered with,
///   or a non-`nil` `expectedFileId` doesn't match the header's embedded id),
///   `.truncated` (the file is shorter than the header requires),
///   `.unsupportedFormatVersion`/`.unsupportedCipher` (the header names one this build doesn't understand),
///   `.malformedHeader` (a header field is structurally invalid, e.g. chunkSize <= 0).
final class VaultFileReader {

    private let fileHandle: FileHandle

    let fileId: Data
    let chunkSize: Int
    let plainSize: Int64
    private let formatVersion: UInt8
    private let cipherId: UInt8
    // `private(set)`, not `private` — external read access (write stays
    // internal-only) is what lets `VaultFileReaderTests.
    // testCloseZeroesFileContentKeyAndCachedChunk` actually observe that
    // `close()` zeroed these in place rather than a copy, the same
    // "otherwise no black-box way to tell zeroed-in-place from zeroed-a-copy"
    // reasoning `VaultStore.create`'s `onCreateFailureVmkForTesting` hook's
    // own doc comment gives for its identical problem.
    private(set) var fileContentKey: Data

    private var cachedChunkIndex: Int64 = -1
    private(set) var cachedChunk: Data?

    /// Counts actual chunk decryptions (cache misses) - lets tests assert the
    /// cache and lazy random-access behavior actually work as designed.
    private(set) var decryptCount: Int = 0

    private var chunkCount: Int64 {
        ChunkedVaultWriter.chunkCountFor(totalPlaintextLength: plainSize, chunkSize: chunkSize)
    }

    /// - Parameter expectedFileId: cross-checked against the fileId embedded
    ///   in the file's own header, closing "this is the wrong file's blob"
    ///   fast rather than surfacing it as an opaque authentication failure
    ///   three steps later on the first chunk read. Pass `nil` only when the
    ///   caller genuinely has no external fileId to check against — the one
    ///   caller that does this today is `VaultManifest.read` (#302), because
    ///   the manifest's fileId varies on every write (see that type's doc
    ///   comment) and is trusted from its own AEAD-authenticated header
    ///   instead. Every other caller must keep passing a concrete id: `nil`
    ///   is not "skip validation," it's "there is provably nothing else to
    ///   validate against."
    init(fileURL: URL, vmk: Data, expectedFileId: Data?) throws {
        precondition(
            expectedFileId == nil || expectedFileId!.count == VaultFormat.fileIdSizeBytes,
            "expectedFileId must be \(VaultFormat.fileIdSizeBytes) bytes"
        )

        let handle = try FileHandle(forReadingFrom: fileURL)

        do {
            let headerData = (try handle.read(upToCount: VaultFormat.headerSizeBytes)) ?? Data()
            guard headerData.count == VaultFormat.headerSizeBytes else {
                throw VaultCryptoError.truncated(
                    expectedBytes: Int64(VaultFormat.headerSizeBytes),
                    actualBytes: Int64(headerData.count)
                )
            }
            let header = [UInt8](headerData)

            let version = header[0]
            let parsedCipherId = header[1]
            let parsedFileId = Data(header[2..<(2 + VaultFormat.fileIdSizeBytes)])
            let parsedChunkSize = BigEndian.int32(header, at: 2 + VaultFormat.fileIdSizeBytes)
            let parsedTotalLength = BigEndian.int64(header, at: 2 + VaultFormat.fileIdSizeBytes + 4)

            guard version == VaultFormat.formatVersion else {
                throw VaultCryptoError.unsupportedFormatVersion(found: version)
            }
            guard parsedCipherId == VaultFormat.cipherAes256Gcm else {
                throw VaultCryptoError.unsupportedCipher(found: parsedCipherId)
            }
            // Not itself an AEAD failure - a fast, clear error for "this is the wrong
            // file's blob" rather than letting the caller discover it as an opaque
            // authentication failure three steps later on the first chunk read.
            // Skipped entirely when expectedFileId is nil - see the init's doc comment.
            if let expectedFileId, expectedFileId != parsedFileId {
                throw VaultCryptoError.authenticationFailed
            }
            // Structural validation BEFORE any arithmetic uses these values - a
            // corrupted chunkSize == 0 must fail cleanly, not crash on a division.
            guard parsedChunkSize > 0 && parsedChunkSize <= VaultFormat.maxReasonableChunkSize else {
                throw VaultCryptoError.malformedHeader("Invalid chunkSize in header: \(parsedChunkSize)")
            }
            guard parsedTotalLength >= 0 else {
                throw VaultCryptoError.malformedHeader("Negative totalPlaintextLength in header: \(parsedTotalLength)")
            }

            self.fileHandle = handle
            self.fileId = parsedFileId
            self.formatVersion = version
            self.cipherId = parsedCipherId
            self.chunkSize = Int(parsedChunkSize)
            self.plainSize = parsedTotalLength
            self.fileContentKey = deriveFileContentKey(vmk: vmk, fileId: parsedFileId)

            // Eagerly authenticate chunk 0 at open time, even for a legitimately empty
            // (plainSize == 0) file. Deliberate, not incidental: readAt() below never
            // touches disk for an empty file (there is nothing to return), which means
            // without this eager check, an attacker could truncate a NON-empty file
            // down to just the header and rewrite totalPlaintextLength to 0 - "erasing"
            // its content as a silent, plausible-looking empty file instead of a loud
            // failure. Decrypting chunk 0 here closes that gap.
            _ = try decryptChunk(0)
        } catch {
            try? handle.close()
            throw error
        }
    }

    private func decryptChunk(_ index: Int64) throws -> Data {
        if let cached = cachedChunk, cachedChunkIndex == index {
            return cached
        }

        let isFinal = index == chunkCount - 1
        let plainLenOfChunk = isFinal ? (plainSize - index * Int64(chunkSize)) : Int64(chunkSize)
        let storedLen = Int(plainLenOfChunk + Int64(VaultFormat.tagSizeBytes))
        let filePos = UInt64(VaultFormat.headerSizeBytes) + UInt64(index) * UInt64(chunkSize + VaultFormat.tagSizeBytes)

        try fileHandle.seek(toOffset: filePos)
        var ciphertext = Data(capacity: storedLen)
        while ciphertext.count < storedLen {
            guard let chunk = try fileHandle.read(upToCount: storedLen - ciphertext.count), !chunk.isEmpty else {
                throw VaultCryptoError.truncated(expectedBytes: Int64(storedLen), actualBytes: Int64(ciphertext.count))
            }
            ciphertext.append(chunk)
        }

        let aad = VaultFormat.chunkAad(
            formatVersion: formatVersion,
            cipherId: cipherId,
            fileId: fileId,
            totalPlaintextLength: plainSize,
            chunkSize: Int32(chunkSize),
            chunkIndex: index,
            isFinalChunk: isFinal
        )
        let nonce = deriveNonce(fileContentKey: fileContentKey, chunkIndex: index)
        let plaintext = try AesGcmCipher.decrypt(key: fileContentKey, nonce: nonce, aad: aad, ciphertextWithTag: ciphertext)

        decryptCount += 1
        cachedChunkIndex = index
        cachedChunk = plaintext
        return plaintext
    }

    /// Reads up to `length` plaintext bytes starting at plaintext `offset`.
    /// Returns fewer bytes than requested only at end-of-file (mirrors the
    /// contract a data-source-style adapter expects).
    ///
    /// - Throws: `.authenticationFailed` (wrong key, or the file was tampered with),
    ///   `.truncated` (fewer bytes exist on disk than the header implies).
    func readAt(offset: Int64, length: Int) throws -> Data {
        guard offset >= 0, length >= 0 else { return Data() }
        guard offset < plainSize else { return Data() }

        let want = Int(min(Int64(length), plainSize - offset))
        var result = Data(capacity: want)
        var produced = 0
        var pos = offset
        while produced < want {
            let index = pos / Int64(chunkSize)
            let withinChunk = Int(pos % Int64(chunkSize))
            let chunk = try decryptChunk(index)
            let take = min(chunk.count - withinChunk, want - produced)
            if take <= 0 { break }
            result.append(chunk[(chunk.startIndex + withinChunk)..<(chunk.startIndex + withinChunk + take)])
            produced += take
            pos += Int64(take)
        }
        return result
    }

    /// Closes the file handle and scrubs this instance's own derived key
    /// material and last-decrypted plaintext chunk — a direct Swift port of
    /// the Kotlin fix from #525/#526 (PR #539's `VaultFileReader.close()`
    /// change), which found the same gap: closing only the file handle left
    /// a live, unscrubbed copy of `fileContentKey` sitting in memory for as
    /// long as this instance itself lived. Idempotent — re-zeroing already-
    /// zeroed `Data` is harmless, so a caller (or `deinit`) calling this
    /// twice is safe.
    func close() {
        try? fileHandle.close()
        fileContentKey.secureZero()
        cachedChunk?.secureZero()
        cachedChunk = nil
    }

    deinit {
        close()
    }
}
